In [68]:
# This Python 3 environment comes with many helpful analytics libraries installed
# It is defined by the kaggle/python Docker image: https://github.com/kaggle/docker-python
# For example, here's several helpful packages to load

import numpy as np # linear algebra
import pandas as pd # data processing, CSV file I/O (e.g. pd.read_csv)

# Input data files are available in the read-only "../input/" directory
# For example, running this (by clicking run or pressing Shift+Enter) will list all files under the input directory

import os
for dirname, _, filenames in os.walk('/kaggle/input'):
    for filename in filenames:
        print(os.path.join(dirname, filename))

# You can write up to 20GB to the current directory (/kaggle/working/) that gets preserved as output when you create a version using "Save & Run All" 
# You can also write temporary files to /kaggle/temp/, but they won't be saved outside of the current session

/kaggle/input/szeged-weather/weatherHistory.csv


# Content
## The CSV file includes a hourly/daily summary for Szeged, Hungary area, between 2006 and 2016.

Data available in the hourly response:

* time
* summary
* precipType
* temperature
* apparentTemperature
* humidity
* windSpeed
* windBearing
* visibility
* loudCover
* pressure

# Step1 : Load the dataset

In [69]:
df = pd.read_csv("../input/szeged-weather/weatherHistory.csv")
df.head(5)

Formatted Date        Summary Precip Type  Temperature (C)  \
0  2006-04-01 00:00:00.000 +0200  Partly Cloudy        rain         9.472222   
1  2006-04-01 01:00:00.000 +0200  Partly Cloudy        rain         9.355556   
2  2006-04-01 02:00:00.000 +0200  Mostly Cloudy        rain         9.377778   
3  2006-04-01 03:00:00.000 +0200  Partly Cloudy        rain         8.288889   
4  2006-04-01 04:00:00.000 +0200  Mostly Cloudy        rain         8.755556   

   Apparent Temperature (C)  Humidity  Wind Speed (km/h)  \
0                  7.388889      0.89            14.1197   
1                  7.227778      0.86            14.2646   
2                  9.377778      0.89             3.9284   
3                  5.944444      0.83            14.1036   
4                  6.977778      0.83            11.0446   

   Wind Bearing (degrees)  Visibility (km)  Loud Cover  Pressure (millibars)  \
0                   251.0          15.8263         0.0               1015.13   
1                   259.0          15.8263         0.0               1015.63   
2                   204.0          14.9569         0.0               1015.94   
3                   269.0          15.8263         0.0               1016.41   
4                   259.0          15.8263         0.0               1016.51   

                       Daily Summary  
0  Partly cloudy throughout the day.  
1  Partly cloudy throughout the day.  
2  Partly cloudy throughout the day.  
3  Partly cloudy throughout the day.  
4  Partly cloudy throughout the day.

In [70]:
df.describe()

Temperature (C)  Apparent Temperature (C)      Humidity  \
count     96453.000000              96453.000000  96453.000000   
mean         11.932678                 10.855029      0.734899   
std           9.551546                 10.696847      0.195473   
min         -21.822222                -27.716667      0.000000   
25%           4.688889                  2.311111      0.600000   
50%          12.000000                 12.000000      0.780000   
75%          18.838889                 18.838889      0.890000   
max          39.905556                 39.344444      1.000000   

       Wind Speed (km/h)  Wind Bearing (degrees)  Visibility (km)  Loud Cover  \
count       96453.000000            96453.000000     96453.000000     96453.0   
mean           10.810640              187.509232        10.347325         0.0   
std             6.913571              107.383428         4.192123         0.0   
min             0.000000                0.000000         0.000000         0.0   
25%             5.828200              116.000000         8.339800         0.0   
50%             9.965900              180.000000        10.046400         0.0   
75%            14.135800              290.000000        14.812000         0.0   
max            63.852600              359.000000        16.100000         0.0   

       Pressure (millibars)  
count          96453.000000  
mean            1003.235956  
std              116.969906  
min                0.000000  
25%             1011.900000  
50%             1016.450000  
75%             1021.090000  
max             1046.380000

In [71]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 96453 entries, 0 to 96452
Data columns (total 12 columns):
 #   Column                    Non-Null Count  Dtype  
---  ------                    --------------  -----  
 0   Formatted Date            96453 non-null  object 
 1   Summary                   96453 non-null  object 
 2   Precip Type               95936 non-null  object 
 3   Temperature (C)           96453 non-null  float64
 4   Apparent Temperature (C)  96453 non-null  float64
 5   Humidity                  96453 non-null  float64
 6   Wind Speed (km/h)         96453 non-null  float64
 7   Wind Bearing (degrees)    96453 non-null  float64
 8   Visibility (km)           96453 non-null  float64
 9   Loud Cover                96453 non-null  float64
 10  Pressure (millibars)      96453 non-null  float64
 11  Daily Summary             96453 non-null  object 
dtypes: float64(8), object(4)
memory usage: 8.8+ MB


In [72]:
df.shape

(96453, 12)

## Summary of Dataset:

* The dataset has 12 columns and 96453 rows
* No null values in the dataset
* Categorical Columns - Formatted Date, Summary, Precip Type, Daily Summary
* Loud Cover only has zero values 
* Increase and Wind Speed and Wind Bearing causes decrease in Apparent Temperature

# Step 2: Data Preprocessing

* Drop Loud Cover and Daily Summary

In [73]:
df.drop(['Loud Cover', 'Daily Summary'], axis=1)

Formatted Date        Summary Precip Type  \
0      2006-04-01 00:00:00.000 +0200  Partly Cloudy        rain   
1      2006-04-01 01:00:00.000 +0200  Partly Cloudy        rain   
2      2006-04-01 02:00:00.000 +0200  Mostly Cloudy        rain   
3      2006-04-01 03:00:00.000 +0200  Partly Cloudy        rain   
4      2006-04-01 04:00:00.000 +0200  Mostly Cloudy        rain   
...                              ...            ...         ...   
96448  2016-09-09 19:00:00.000 +0200  Partly Cloudy        rain   
96449  2016-09-09 20:00:00.000 +0200  Partly Cloudy        rain   
96450  2016-09-09 21:00:00.000 +0200  Partly Cloudy        rain   
96451  2016-09-09 22:00:00.000 +0200  Partly Cloudy        rain   
96452  2016-09-09 23:00:00.000 +0200  Partly Cloudy        rain   

       Temperature (C)  Apparent Temperature (C)  Humidity  Wind Speed (km/h)  \
0             9.472222                  7.388889      0.89            14.1197   
1             9.355556                  7.227778      0.86            14.2646   
2             9.377778                  9.377778      0.89             3.9284   
3             8.288889                  5.944444      0.83            14.1036   
4             8.755556                  6.977778      0.83            11.0446   
...                ...                       ...       ...                ...   
96448        26.016667                 26.016667      0.43            10.9963   
96449        24.583333                 24.583333      0.48            10.0947   
96450        22.038889                 22.038889      0.56             8.9838   
96451        21.522222                 21.522222      0.60            10.5294   
96452        20.438889                 20.438889      0.61             5.8765   

       Wind Bearing (degrees)  Visibility (km)  Pressure (millibars)  
0                       251.0          15.8263               1015.13  
1                       259.0          15.8263               1015.63  
2                       204.0          14.9569               1015.94  
3                       269.0          15.8263               1016.41  
4                       259.0          15.8263               1016.51  
...                       ...              ...                   ...  
96448                    31.0          16.1000               1014.36  
96449                    20.0          15.5526               1015.16  
96450                    30.0          16.1000               1015.66  
96451                    20.0          16.1000               1015.95  
96452                    39.0          15.5204               1016.16  

[96453 rows x 10 columns]

## Formatted Date currently has object dtype, convert into datetime format

In [74]:
import datetime

df['Formatted Date'] = pd.to_datetime(df['Formatted Date'], utc=True)
df['Formatted Date'].sample(5)

56849   2012-07-31 15:00:00+00:00
82434   2015-01-05 20:00:00+00:00
73526   2014-01-28 13:00:00+00:00
55026   2012-02-17 17:00:00+00:00
43379   2010-09-19 09:00:00+00:00
Name: Formatted Date, dtype: datetime64[ns, UTC]

In [75]:
df['year'] = df['Formatted Date'].dt.year
df['month'] = df['Formatted Date'].dt.month
df['day'] = df['Formatted Date'].dt.day
df['weekday'] = df['Formatted Date'].dt.weekday
df.head(5)

Formatted Date        Summary Precip Type  Temperature (C)  \
0 2006-03-31 22:00:00+00:00  Partly Cloudy        rain         9.472222   
1 2006-03-31 23:00:00+00:00  Partly Cloudy        rain         9.355556   
2 2006-04-01 00:00:00+00:00  Mostly Cloudy        rain         9.377778   
3 2006-04-01 01:00:00+00:00  Partly Cloudy        rain         8.288889   
4 2006-04-01 02:00:00+00:00  Mostly Cloudy        rain         8.755556   

   Apparent Temperature (C)  Humidity  Wind Speed (km/h)  \
0                  7.388889      0.89            14.1197   
1                  7.227778      0.86            14.2646   
2                  9.377778      0.89             3.9284   
3                  5.944444      0.83            14.1036   
4                  6.977778      0.83            11.0446   

   Wind Bearing (degrees)  Visibility (km)  Loud Cover  Pressure (millibars)  \
0                   251.0          15.8263         0.0               1015.13   
1                   259.0          15.8263         0.0               1015.63   
2                   204.0          14.9569         0.0               1015.94   
3                   269.0          15.8263         0.0               1016.41   
4                   259.0          15.8263         0.0               1016.51   

                       Daily Summary  year  month  day  weekday  
0  Partly cloudy throughout the day.  2006      3   31        4  
1  Partly cloudy throughout the day.  2006      3   31        4  
2  Partly cloudy throughout the day.  2006      4    1        5  
3  Partly cloudy throughout the day.  2006      4    1        5  
4  Partly cloudy throughout the day.  2006      4    1        5

In [76]:
df = df.drop(['Loud Cover', 'Daily Summary'], axis=1)
df.head(5)

Formatted Date        Summary Precip Type  Temperature (C)  \
0 2006-03-31 22:00:00+00:00  Partly Cloudy        rain         9.472222   
1 2006-03-31 23:00:00+00:00  Partly Cloudy        rain         9.355556   
2 2006-04-01 00:00:00+00:00  Mostly Cloudy        rain         9.377778   
3 2006-04-01 01:00:00+00:00  Partly Cloudy        rain         8.288889   
4 2006-04-01 02:00:00+00:00  Mostly Cloudy        rain         8.755556   

   Apparent Temperature (C)  Humidity  Wind Speed (km/h)  \
0                  7.388889      0.89            14.1197   
1                  7.227778      0.86            14.2646   
2                  9.377778      0.89             3.9284   
3                  5.944444      0.83            14.1036   
4                  6.977778      0.83            11.0446   

   Wind Bearing (degrees)  Visibility (km)  Pressure (millibars)  year  month  \
0                   251.0          15.8263               1015.13  2006      3   
1                   259.0          15.8263               1015.63  2006      3   
2                   204.0          14.9569               1015.94  2006      4   
3                   269.0          15.8263               1016.41  2006      4   
4                   259.0          15.8263               1016.51  2006      4   

   day  weekday  
0   31        4  
1   31        4  
2    1        5  
3    1        5  
4    1        5

# Step-3: Create Pipeline for Category Encoding and Model Predictions

In [77]:
df['Precip Type']= df['Precip Type'].dropna()
df['Precip Type'].unique()

array(['rain', 'snow', nan], dtype=object)

In [78]:
df['Precip Type'] = df['Precip Type'].fillna("None")
df["Precip Type"] = df['Precip Type'].astype('str')
df['Summary'] = df['Summary'].astype('str')
df["Precip Type"].dtype, df["Summary"].dtype

(dtype('O'), dtype('O'))

In [80]:
from sklearn.preprocessing import OrdinalEncoder
from category_encoders.james_stein import JamesSteinEncoder

In [81]:
ordinal_columns = ['Precip Type', 'Summary']
#jstein_columns = ['Summary']

In [82]:
from sklearn.compose import ColumnTransformer
from sklearn.pipeline import Pipeline
from sklearn.linear_model import LinearRegression
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import LabelEncoder

preprocessor = ColumnTransformer(
    transformers=[
        ('ord', OrdinalEncoder(), ordinal_columns),
        #('cat', JamesSteinEncoder(), jstein_columns)
    ])

model = LinearRegression()

pro_pipeline = Pipeline([('preprocessor', preprocessor),('model', model)])

In [100]:
X = df.drop(['Apparent Temperature (C)'], axis=1)
X= X.drop(['Formatted Date'], axis=1)
y = df['Apparent Temperature (C)']
X.head(5)

Summary Precip Type  Temperature (C)  Humidity  Wind Speed (km/h)  \
0  Partly Cloudy        rain         9.472222      0.89            14.1197   
1  Partly Cloudy        rain         9.355556      0.86            14.2646   
2  Mostly Cloudy        rain         9.377778      0.89             3.9284   
3  Partly Cloudy        rain         8.288889      0.83            14.1036   
4  Mostly Cloudy        rain         8.755556      0.83            11.0446   

   Wind Bearing (degrees)  Visibility (km)  Pressure (millibars)  year  month  \
0                   251.0          15.8263               1015.13  2006      3   
1                   259.0          15.8263               1015.63  2006      3   
2                   204.0          14.9569               1015.94  2006      4   
3                   269.0          15.8263               1016.41  2006      4   
4                   259.0          15.8263               1016.51  2006      4   

   day  weekday  
0   31        4  
1   31        4  
2    1        5  
3    1        5  
4    1        5

In [121]:
cat_encoder = LabelEncoder()
ord_encoder = OrdinalEncoder()

X['enc_summary'] = cat_encoder.fit_transform(X['Summary'])
X['ord_summary'] = ord_encoder.fit_transform(X['Summary'].values.reshape(-1,1))
X['ord_precip'] = ord_encoder.fit_transform(X['Precip Type'].values.reshape(-1,1))
X = X.drop(['Summary', 'Precip Type', 'enc_summary'], axis=1)
X.head(5)

KeyError: 'Summary'

0    7.388889
1    7.227778
2    9.377778
3    5.944444
4    6.977778
Name: Apparent Temperature (C), dtype: float64

In [116]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.25, random_state=0)

In [117]:
X_train.shape, y_train.shape

((72339, 12), (72339,))

In [118]:
model = LinearRegression()

model.fit(X_train, y_train)

y_pred = model.predict(X_test)

In [122]:
from sklearn.metrics import r2_score

score = r2_score(y_test, y_pred)
print(score)

0.9899363541105147
